s3/s3_fs.go

```diff
@@ -0,0 +1,235 @@
+}
+
+// OpenFile opens a file.
+func (fs S3Fs) OpenFile(name string, flag int, perm os.FileMode) (afero.File, error) {
+	file := NewS3File(fs.bucket, name, fs.s3API, fs)
+	if flag&os.O_APPEND != 0 {
+		return file, errors.New("S3 is eventually consistent. Appending files will lead to trouble.")
+	}
+	if flag&os.O_CREATE != 0 {
+		if _, err := file.WriteString(""); err != nil {
+			return file, err
+		}
+	}
+	return file, nil
+}
+
+// Remove a file.
+func (fs S3Fs) Remove(name string) error {
+	if _, err := fs.Stat(name); err != nil {
+		return err
+	}
+	_, err := fs.s3API.DeleteObject(&s3.DeleteObjectInput{
+		Bucket: aws.String(fs.bucket),
+		Key:    aws.String(name),
+	})
+	return err
+}
+
+// ForceRemove doesn't error if a file does not exist.
+func (fs S3Fs) ForceRemove(name string) error {
+	_, err := fs.s3API.DeleteObject(&s3.DeleteObjectInput{
+		Bucket: aws.String(fs.bucket),
+		Key:    aws.String(name),
+	})
+	return err
+}
+
+// RemoveAll removes a path.
+func (fs S3Fs) RemoveAll(path string) error {
+	s3dir := NewS3File(fs.bucket, path, fs.s3API, fs)
+	fis, err := s3dir.Readdir(0)
+	if err != nil {
+		return err
+	}
+	for _, fi := range fis {
+		fullpath := filepath.Join(s3dir.Name(), fi.Name())
+		if fi.IsDir() {
+			if err := fs.RemoveAll(fullpath); err != nil {
+				return err
+			}
+		} else {
+			if err := fs.ForceRemove(fullpath); err != nil {
+				return err
+			}
+		}
+	}
+	// finally remove the "file" representing the directory
+	if err := fs.ForceRemove(s3dir.Name() + "/"); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Rename a file.
+// There is no method to directly rename an S3 object, so the Rename
+// will copy the file to an object with the new name and then delete
+// the original.
+func (fs S3Fs) Rename(oldname, newname string) error {
+	if oldname == newname {
+		return nil
+	}
+	_, err := fs.s3API.CopyObject(&s3.CopyObjectInput{
+		Bucket:               aws.String(fs.bucket),
+		CopySource:           aws.String(fs.bucket + oldname),
+		Key:                  aws.String(newname),
+		ServerSideEncryption: aws.String("AES256"),
+	})
+	if err != nil {
+		return err
+	}
+	_, err = fs.s3API.DeleteObject(&s3.DeleteObjectInput{
+		Bucket: aws.String(fs.bucket),
+		Key:    aws.String(oldname),
+	})
+	return err
+}
+
+func hasTrailingSlash(s string) bool {
+	return len(s) > 0 && s[len(s)-1] == '/'
+}
+
+func trimLeadingSlash(s string) string {
+	if len(s) > 0 && s[0] == '/' {
+		return s[1:]
+	}
+	return s
+}
+
+// Stat returns a FileInfo describing the named file.
+// If there is an error, it will be of type *os.PathError.
+func (fs S3Fs) Stat(name string) (os.FileInfo, error) {
+	nameClean := filepath.Clean(name)
+	out, err := fs.s3API.HeadObject(&s3.HeadObjectInput{
+		Bucket: aws.String(fs.bucket),
+		Key:    aws.String(nameClean),
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), "404") {
+			statDir, err := fs.statDirectory(name)
+			return statDir, err
+		}
```

Comment:
Rather than munging the string, I think it would be better to check the error code:

```go
if err, ok := err.(awserr.Error); ok {
    if err.Code() == "NotFound" {
        return fs.statDirectory(name)
    }
}
```